src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -378,69 +365,131 @@ and gen_from_arrow ~loc ~env left right =
+
+    - `Recursive of expression * expression
+    The derivated generator was recursive (i.e. val gen : n -> t Gen.t), we return
+    the sized generator version, and a normal generator using this last with
+    [Gen.sized].
+*)
+let gen_from_type_declaration ~loc td =
+  let (module A) = Ast_builder.make loc in
+  let ty = curr_type () in
+  let is_rec = is_rec (curr_type ()) in
 
   let args =
     List.map
       (fun (typ, _) ->
-        match typ.ptyp_desc with Ptyp_var s -> pat ~loc s | _ -> assert false)
+        match typ.ptyp_desc with
+        | Ptyp_var s -> (pat ~loc s, name s |> A.evar)
+        | _ -> assert false)
       td.ptype_params
   in
+  let (args_pat, args_expr) = List.split args in
 
   let gen =
     match td.ptype_kind with
-    | Ptype_variant xs -> gen_from_kind_variant ~loc ~env name xs
+    | Ptype_variant xs -> gen_from_kind_variant ~loc xs
     | Ptype_record xs ->
-        let gens = List.map (fun x -> gen_from_type ~loc ~env x.pld_type) xs in
+        let gens = List.map (fun x -> gen_from_type ~loc  x.pld_type) xs in
         record ~loc ~gens xs
     | _ ->
         let typ = Option.get td.ptype_manifest in
-        gen_from_type ~loc ~env ~typ_name:name typ
+        gen_from_type ~loc  typ
+  in
+
+  let pat_gen = pat ~loc ty in
+  if not is_rec then
+    let gen = Args.curry_args ~loc args_pat gen in
+    `Normal [%stri let [%p pat_gen] = [%e gen]]
+  else
+    let gen = Args.curry_args ~loc (args_pat @ [A.pvar "n"]) gen in
+    let pat_gen_sized = pat ~loc ~sized:true ty in
+    let gen_sized = name ~sized:true ty |> A.evar in
+    let gen_normal =
+      Args.curry_args ~loc args_pat
+        (G.sized ~loc (Args.apply_args ~loc args_expr gen_sized))
+    in
+    `Recursive (
+        [%stri let rec [%p pat_gen_sized] = [%e gen]],
+        [%stri let [%p pat_gen] = [%e gen_normal]]
+      )
+
+(** *)
+let mutually_recursive_gens ~loc gens =
+  let (module A) = Ast_builder.make loc in
+  let to_mutualize_gens =
+    List.map (function
+        | `Recursive (x, _) -> x
+        | `Normal x -> x) gens
+  in
+  let normal_gens =
+    List.filter_map (function
+        | `Recursive (_, x) -> Some x
+        | `Normal _ -> None) gens
   in
-  let gen = curry_args ~loc args gen in
+  let gens =
+    List.map (function
+        | [%stri let [%p? pat] = [%e? expr]]
+          | [%stri let rec [%p? pat] = [%e? expr]] ->
+           A.value_binding ~pat ~expr
+        | _ -> assert false) to_mutualize_gens
+  in
+  let mutual_gens = A.pstr_value Recursive gens in
+  mutual_gens :: normal_gens
 
-  [%stri let [%p pat_gen] = [%e gen]]
+(** [derive_gen ~loc xs] creates generators for type declaration in [xs].
 
-let derive_gen ~loc xs : structure =
-  match xs with
-  | _, [ x ] -> [ gen_from_type_declaration ~loc x ]
+    He also has an hidden purpose: he sets the environment prior to
+    the derivation. He identifies recursive types declarations (e.g list)
+    and its external arguments (e.g. 'a list).
```

Comment:
Minor (lang): *It* also ... : *it* sets ... *It* identifies  (I'm not a native English speaker either 🙂)

Reply:
Fixed in 696b9b9aecbcdeaa75b1324f2ab75f0f813630f2.